// fortune-lib/lib/enum_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * 地支
 */
var EARTH_SYMBOL;
(function (EARTH_SYMBOL) {
    EARTH_SYMBOL[EARTH_SYMBOL["\u5B50"] = 0] = "\u5B50";
    EARTH_SYMBOL[EARTH_SYMBOL["\u4E11"] = 1] = "\u4E11";
    EARTH_SYMBOL[EARTH_SYMBOL["\u5BC5"] = 2] = "\u5BC5";
    EARTH_SYMBOL[EARTH_SYMBOL["\u536F"] = 3] = "\u536F";
    EARTH_SYMBOL[EARTH_SYMBOL["\u8FB0"] = 4] = "\u8FB0";
    EARTH_SYMBOL[EARTH_SYMBOL["\u5DF3"] = 5] = "\u5DF3";
    EARTH_SYMBOL[EARTH_SYMBOL["\u5348"] = 6] = "\u5348";
    EARTH_SYMBOL[EARTH_SYMBOL["\u672A"] = 7] = "\u672A";
    EARTH_SYMBOL[EARTH_SYMBOL["\u7533"] = 8] = "\u7533";
    EARTH_SYMBOL[EARTH_SYMBOL["\u9149"] = 9] = "\u9149";
    EARTH_SYMBOL[EARTH_SYMBOL["\u620C"] = 10] = "\u620C";
    EARTH_SYMBOL[EARTH_SYMBOL["\u4EA5"] = 11] = "\u4EA5";
})(EARTH_SYMBOL || (EARTH_SYMBOL = {}));
exports.EARTH_SYMBOL = EARTH_SYMBOL;
;
/**
 * 六親
 */
var SIX_RELATIVE_SYMBOL;
(function (SIX_RELATIVE_SYMBOL) {
    SIX_RELATIVE_SYMBOL[SIX_RELATIVE_SYMBOL["\u7236"] = 0] = "\u7236";
    SIX_RELATIVE_SYMBOL[SIX_RELATIVE_SYMBOL["\u5144"] = 1] = "\u5144";
    SIX_RELATIVE_SYMBOL[SIX_RELATIVE_SYMBOL["\u5B50"] = 2] = "\u5B50";
    SIX_RELATIVE_SYMBOL[SIX_RELATIVE_SYMBOL["\u624D"] = 3] = "\u624D";
    SIX_RELATIVE_SYMBOL[SIX_RELATIVE_SYMBOL["\u5B98"] = 4] = "\u5B98";
})(SIX_RELATIVE_SYMBOL || (SIX_RELATIVE_SYMBOL = {}));
exports.SIX_RELATIVE_SYMBOL = SIX_RELATIVE_SYMBOL;
;
/**
 * 天干
 */
var HEAVEN_SYMBOL;
(function (HEAVEN_SYMBOL) {
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u7532"] = 0] = "\u7532";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u4E59"] = 1] = "\u4E59";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u4E19"] = 2] = "\u4E19";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u4E01"] = 3] = "\u4E01";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u620A"] = 4] = "\u620A";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u5DF1"] = 5] = "\u5DF1";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u5E9A"] = 6] = "\u5E9A";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u8F9B"] = 7] = "\u8F9B";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u58EC"] = 8] = "\u58EC";
    HEAVEN_SYMBOL[HEAVEN_SYMBOL["\u7678"] = 9] = "\u7678";
})(HEAVEN_SYMBOL || (HEAVEN_SYMBOL = {}));
exports.HEAVEN_SYMBOL = HEAVEN_SYMBOL;
;
/**
 * 五行
 */
var FIVE_ELEMENT;
(function (FIVE_ELEMENT) {
    FIVE_ELEMENT[FIVE_ELEMENT["\u91D1"] = 0] = "\u91D1";
    FIVE_ELEMENT[FIVE_ELEMENT["\u6C34"] = 1] = "\u6C34";
    FIVE_ELEMENT[FIVE_ELEMENT["\u6728"] = 2] = "\u6728";
    FIVE_ELEMENT[FIVE_ELEMENT["\u706B"] = 3] = "\u706B";
    FIVE_ELEMENT[FIVE_ELEMENT["\u571F"] = 4] = "\u571F";
})(FIVE_ELEMENT || (FIVE_ELEMENT = {}));
exports.FIVE_ELEMENT = FIVE_ELEMENT;
;
/**
 * 順向 上掛推定
 */
var EARTH_SYMBOL_ORDER;
(function (EARTH_SYMBOL_ORDER) {
    EARTH_SYMBOL_ORDER[EARTH_SYMBOL_ORDER["\u5B50"] = 0] = "\u5B50";
    EARTH_SYMBOL_ORDER[EARTH_SYMBOL_ORDER["\u5BC5"] = 1] = "\u5BC5";
    EARTH_SYMBOL_ORDER[EARTH_SYMBOL_ORDER["\u8FB0"] = 2] = "\u8FB0";
    EARTH_SYMBOL_ORDER[EARTH_SYMBOL_ORDER["\u5348"] = 3] = "\u5348";
    EARTH_SYMBOL_ORDER[EARTH_SYMBOL_ORDER["\u7533"] = 4] = "\u7533";
    EARTH_SYMBOL_ORDER[EARTH_SYMBOL_ORDER["\u620C"] = 5] = "\u620C";
})(EARTH_SYMBOL_ORDER || (EARTH_SYMBOL_ORDER = {}));
exports.EARTH_SYMBOL_ORDER = EARTH_SYMBOL_ORDER;
;
/**
 * 逆向 上掛推定
 */
var EARTH_SYMBOL_REVERSE;
(function (EARTH_SYMBOL_REVERSE) {
    EARTH_SYMBOL_REVERSE[EARTH_SYMBOL_REVERSE["\u672A"] = 0] = "\u672A";
    EARTH_SYMBOL_REVERSE[EARTH_SYMBOL_REVERSE["\u5DF3"] = 1] = "\u5DF3";
    EARTH_SYMBOL_REVERSE[EARTH_SYMBOL_REVERSE["\u536F"] = 2] = "\u536F";
    EARTH_SYMBOL_REVERSE[EARTH_SYMBOL_REVERSE["\u4E11"] = 3] = "\u4E11";
    EARTH_SYMBOL_REVERSE[EARTH_SYMBOL_REVERSE["\u4EA5"] = 4] = "\u4EA5";
    EARTH_SYMBOL_REVERSE[EARTH_SYMBOL_REVERSE["\u9149"] = 5] = "\u9149";
})(EARTH_SYMBOL_REVERSE || (EARTH_SYMBOL_REVERSE = {}));
exports.EARTH_SYMBOL_REVERSE = EARTH_SYMBOL_REVERSE;
;
/**
 * 六獸
 */
var SIX_ANIMALS;
(function (SIX_ANIMALS) {
    SIX_ANIMALS[SIX_ANIMALS["\u9752\u9F8D"] = 0] = "\u9752\u9F8D";
    SIX_ANIMALS[SIX_ANIMALS["\u6731\u96C0"] = 1] = "\u6731\u96C0";
    SIX_ANIMALS[SIX_ANIMALS["\u52FE\u5875"] = 2] = "\u52FE\u5875";
    SIX_ANIMALS[SIX_ANIMALS["\u9A30\u86C7"] = 3] = "\u9A30\u86C7";
    SIX_ANIMALS[SIX_ANIMALS["\u767D\u864E"] = 4] = "\u767D\u864E";
    SIX_ANIMALS[SIX_ANIMALS["\u7384\u6B66"] = 5] = "\u7384\u6B66";
})(SIX_ANIMALS || (SIX_ANIMALS = {}));
exports.SIX_ANIMALS = SIX_ANIMALS;
